docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,216 @@
+```
+
+The output becomes:
+```text
+group: a
+   recording:<name:"r1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<seconds:2 > > 
+```
+
+Given the following stream of incoming alerting rules:
+```text
+group: a
+   alert:<state:FIRING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<seconds:1 > > 
+group: a
+   alert:<             name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-2" > > last_evaluation:<seconds:2 > > 
+```
+
+The output becomes:
+```text
+group: a
+   alert:<state:FIRING name:"a1" labels:<labels:<name:"replica" value:"thanos-ruler-1" > > last_evaluation:<seconds:1 > > 
+```
+
+Note how in the above output the firing alerting rule was preferred despite being older.
+
+Scenario 3:
+
+If, under the above conditions a rule is a candidate for deduplication, finally the rule `expr` and `for` fields are being considered for deduplication.
+
+Given the following stream of incoming alerting rules will also result in two independent alerting rules as both the `expr` and `for` fields differ:
+```text
+  - alert: KubeAPIErrorBudgetBurn
+    annotations:
+      message: The API server is burning too much error budget
+    expr: |
+      sum(apiserver_request:burnrate1h) > (14.40 * 0.01000)
+      and
+      sum(apiserver_request:burnrate5m) > (14.40 * 0.01000)
+    for: 2m
+    labels:
+      severity: critical
+  - alert: KubeAPIErrorBudgetBurn
+    annotations:
+      message: The API server is burning too much error budget
+    expr: |
+      sum(apiserver_request:burnrate6h) > (6.00 * 0.01000)
+      and
+      sum(apiserver_request:burnrate30m) > (6.00 * 0.01000)
+    for: 15m
+    labels:
+      severity: critical
+```
+
+### Type and Info
+
+To accommodate for future additional APIs alongside with the core Rules proto changes an enhancement of the protobuf Info and Type structures are suggested.
+
+ The current `StoreType` enum is renamed to `Type`. This retains binary compatibility with older clients:
+
+ ```diff
+-enum StoreType {
++enum Type {
+   UNKNOWN = 0;
+   QUERY = 1;
+   RULE = 2;
+```
+
+ The current fiels in `InfoResponse` connected to Store APIs are deprecated and dedicated new API sub types are proposed:
+
+ ```diff
+ message InfoResponse {
+   // Deprecated. Use label_sets instead.
+   repeated Label labels = 1 [(gogoproto.nullable) = false];
++  // Deprecated. Will be removed in favor of StoreInfoResponse in the future.
+   int64 min_time = 2;
++  // Deprecated. Will be removed in favor of StoreInfoResponse in the future.
+   int64 max_time = 3;
+-  StoreType storeType  = 4;
++  Type type  = 4;
+   // label_sets is an unsorted list of `LabelSet`s.
++  // Deprecated. Will be removed in favor of StoreInfoResponse in the future.
+   repeated LabelSet label_sets = 5 [(gogoproto.nullable) = false];
++
++  StoreInfoResponse store = 6;
++  RulesInfoResponse rules = 7;
```

Comment:
What is this? Why is this needed? Where are the protos for this?

Reply:
Agreed, see https://github.com/thanos-io/thanos/pull/2263#discussion_r410143232, let's handle this in a separate proposal.

Reply:
Maybe let's put into # Future so it;s not lost???

Reply:
+1 on future

Reply:
Moved to a `## Future` section at the end :+1: